src/Authenticator/HttpDigestAuthenticator.php

```diff
@@ -199,17 +197,17 @@ public static function password($username, $password, $realm)
     protected function loginHeaders(ServerRequestInterface $request)
     {
         $server = $request->getServerParams();
-        $realm = $this->_config['realm'] ?: $server['SERVER_NAME'];
+        $realm = $this->getConfig('realm') === null ? $server['SERVER_NAME'] : $this->getConfig('realm');
 
         $options = [
             'realm' => $realm,
-            'qop' => $this->_config['qop'],
+            'qop' => $this->getConfig('qop'),
             'nonce' => $this->generateNonce(),
-            'opaque' => $this->_config['opaque'] ?: md5($realm)
+            'opaque' => $this->getConfig('opaque') === null ? md5($realm) : $this->getConfig('opaque'),
         ];
 
         $digest = $this->_getDigest($request);
-        if ($digest && isset($digest['nonce']) && !$this->validNonce($digest['nonce'])) {
+        if ($digest !== null && array_key_exists('nonce', $digest) && !$this->validNonce($digest['nonce'])) {
```

Comment:
Same thing, anything else than isset() is not sane, especially not array_key_exists()

Reply:
Well depends on how precise `_getDigest()` is?